### PR TITLE
atomic registry : restrict views based on roles

### PR DESCRIPTION
**Pull request by sub-mod**

### CODE REVIEW

File: pkg/kubernetes/scripts/projects.js [hunk 1]
```diff
@@ -76,6 +76,11 @@
             loader.watch("users");
             loader.watch("groups");
             loader.watch("policybindings");
+            $scope.currUser = "";
+            loader.load("/oapi/v1/users/~").then(function(response){
```
Comment (petervo) [inline pkg/kubernetes/scripts/projects.js]: What about doing this in the general settings, same place we setup isAdmin

File: pkg/kubernetes/scripts/projects.js [hunk 2]
```diff
@@ -405,6 +425,14 @@
 
                 return false;
             }
+            function isCurrentUser(currUser, user) {
+                var userName = toName(user);
+                var currentUser = false;
```
Comment (petervo) [inline pkg/kubernetes/scripts/projects.js]: can probably be simplified to:
```return currUser ? currUser === userName : false;```

File: pkg/kubernetes/views/project-listing.html [hunk 3]
```diff
@@ -32,6 +32,12 @@ <h3 translatable="yes">Projects</h3>
         </tr>
     </tbody>
     <tbody data-ng-rubbish="" ng-if="0" ng-repeat-end></tbody>
+    <thead class="listing-empty" ng-if="!quiet">
```
Comment (petervo) [inline pkg/kubernetes/views/project-listing.html]: quiet is no longer a thing

**Comment by petervo** [comment]
Are there designs you are pulling from? I'm not sure why some of these elements are only for the current user, but maybe that's the way it's designed.

File: pkg/kubernetes/scripts/projects.js [hunk 4]
```diff
@@ -212,7 +212,7 @@
         'kubeLoader',
         'projectPolicy',
         function(select, loader, policy) {
-            var registryRoles = [{ ocRole: "registry-admin", displayRole :"Admin"},
+            var registryRoles = [{ ocRole: "admin", displayRole :"Admin"},
```
Comment (aweiteka) [inline pkg/kubernetes/scripts/projects.js]: If user's ocRole is "admin" or "registry-admin" we want displayRole: "Admin"

**Comment by aweiteka** [comment]
I like the disabled / grey'ed out enhancement. This PR needs to support both admin and registry-admin as valid roles.

**Comment by stefwalter** [comment]
Tests seem to fail (see rhel-atomic above):

```
Page error: Error: undefined is not an object (evaluating 'n.groups')
```

File: pkg/kubernetes/scripts/kube-client-cockpit.js [hunk 5]
```diff
@@ -1052,8 +1052,14 @@
                     },
                     flavor: "kubernetes",
                     isAdmin: false,
+                    currUser: null,
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/kube-client-cockpit.js]: Lets use the full word 'currentUser' here.

File: pkg/kubernetes/scripts/projects.js [hunk 6]
```diff
@@ -937,14 +966,20 @@
...
-
-                return methods.check(request, { "metadata.name": "#project-new-name" })
-                    .then(function() {
-                        return methods.create(request)
-                            .then(function() {
-                                return projectData.shareImages(name, fields.access === "shared");
-                            });
-                    });
+                var chain = $q.when();
+                chain = chain.then(function() {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: There's no need to continue to assign chain here.

This is enough:

```
return methods.check(request, ...)
    .then(function() {
        return methods.create(request);
    })
    .then(function() {
       return projectData.shareImages...
    })
    ...
}}}

File: pkg/kubernetes/scripts/kube-client-cockpit.js [hunk 7]
```diff
@@ -1052,8 +1052,14 @@
...
                     canChangeConnection: false,
                 };
+                var curr_user = loader.load("/oapi/v1/users/~")
+                    .then(function(response){
+                        if (response)
+                            settings.currUser = response;
+                    }, function(ex) {});
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/kube-client-cockpit.js]: We should only allow 404 to fall through here. Other failures should result in a real failure displayed in the UI.

File: pkg/kubernetes/scripts/projects.js [hunk 8]
```diff
@@ -261,7 +262,8 @@
                 angular.forEach(roleBinds, function(roleBind) {
                     meta = roleBind.metadata || { };
                     if (meta.name && ocRegistryRoles.indexOf(meta.name)!== -1) {
-                        roles.push(getDisplayRole(meta.name));
+                        if(roles.indexOf(getDisplayRole(meta.name)== -1))
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: Needs a triple equals and a space.

File: pkg/kubernetes/scripts/projects.js [hunk 9]
```diff
@@ -406,6 +408,24 @@
                 return false;
             }
 
+            function makeRegistryAdmin(project, currentUser) {
+                var subject = {
+                    kind: 'User',
+                    name: toName(currentUser),
+                };
+                //if system:admin return
+                if(currentUser.groups.indexOf("system:cluster-admins")!= -1) {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: Needs a space after if.

Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: Needs !== and a space

File: pkg/kubernetes/scripts/projects.js [hunk 10]
```diff
@@ -406,6 +408,24 @@
...
+                    var deferred = $q.defer();
+                    deferred.resolve();
+                    return deferred.promise;
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: What are we resolving here. You can just use ```$q.when({ });``` or something like that, instead of these three lines.

File: pkg/kubernetes/scripts/projects.js [hunk 11]
```diff
@@ -937,14 +966,20 @@
...
+                    return methods.check(request, { "metadata.name": "#project-new-name" });
+                });
+                chain = chain.then(function() { 
+                    return methods.create(request);
+                });
+                chain = chain.then(function() { 
+                    return projectData.shareImages(name, fields.access === "shared");
+                });
+                chain = chain.then(function() {
+                    return projectData.makeRegistryAdmin(name, dialogData.currUser);
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: Why is this necessary? I don't see that ```oc new-project``` has to perform this action:

```
oc new-project --loglevel=10 blah
oc get policybindings --output=json
```

The above shows scruffy an admin already without being added manually.

File: pkg/kubernetes/views/project-body.html [hunk 12]
```diff
@@ -26,27 +26,37 @@
...
+                                value="{{ roleMp.displayRole }}">
+                                {{ roleMp.displayRole }}
+                            </a>
+                        </li>
+                    </ul>
+                </div>
+                <div ng-if="isCurrentUser(settings.currUser, user)"  class="btn-group bootstrap-select namespace-filter" dropdown disabled>
+                    <button class="btn btn-default dropdown-toggle" type="button" dropdown-toggle disabled>
+                        <span class="pull-left">
+                            <span translatable="yes">
```
Comment (stefwalter) [inline pkg/kubernetes/views/project-body.html]: Translatable tags should be on one line.

File: pkg/kubernetes/scripts/projects.js [hunk 13]
```diff
@@ -406,6 +408,24 @@
...
+                //if system:admin return
+                if(currentUser.groups.indexOf("system:cluster-admins")!= -1) {
+                    var deferred = $q.defer();
+                    deferred.resolve();
+                    return deferred.promise;
+                } 
+                return policy.addToRole(project, 'registry-admin', subject);
+            }
+
+            function isCurrentUser(currUser, userName) {
```
Comment (stefwalter) [inline pkg/kubernetes/scripts/projects.js]: This should be called ```isSameUser```

File: pkg/kubernetes/scripts/projects.js (hunk 11 above)
Comment (sub-mod) [inline pkg/kubernetes/scripts/projects.js]: @stefwalter @aweiteka 
for time being we need this to make registry-admin useful in our UI

File: pkg/kubernetes/scripts/kube-client-cockpit.js [hunk 14]
```diff
@@ -1052,8 +1052,14 @@
                     },
                     flavor: "kubernetes",
                     isAdmin: false,
+                    currentUser: null,
```
Comment (sub-mod) [inline pkg/kubernetes/scripts/kube-client-cockpit.js]: @stefwalter made this currentUser

File: pkg/kubernetes/scripts/projects.js [hunk 15]
```diff
@@ -940,10 +977,13 @@
...
-                            .then(function() {
-                                return projectData.shareImages(name, fields.access === "shared");
-                            });
+                        return methods.create(request);
+                    })
+                    .then(function(){
+                        return projectData.shareImages(name, fields.access === "shared");
+                    })
+                    .then(function(){
+                        return projectData.makeRegistryAdmin(name, dialogData.currUser);
```
Comment (sub-mod) [inline pkg/kubernetes/scripts/projects.js]: @stefwalter @aweiteka
for time being we need this to make registry-admin useful in our UI

@aweiteka IMO we need to add a project template which will not create a 'admin' policy. ok  ?

Comment (aweiteka) [inline pkg/kubernetes/scripts/projects.js]: > for time being we need this to make registry-admin useful in our UI
> 
> @aweiteka IMO we need to add a project template which will not create a 'admin' policy. ok ?

At this point I'm much less concerned about if the user is admin or registry-admin. The UI should reflect that they are "Admin" either way. Also, we want them to be assigned role "registry-admin" role when new admin users are added to a project. That's really it. Then it will work regardless of  the project template we use.

Comment (sub-mod) [inline pkg/kubernetes/scripts/projects.js]: As mentioned in irc. 
'admin' user is created by default template.'admin'  user should not be used in registry UI since we have registry-admin for Admin related roles.
makeRegistryAdmin will enforce this even when default template is used.
There is ofcourse no way to add/remove 'admin' role from registry UI.Which makes this safe.

**Comment by sub-mod** [comment]
All the tescases in check-openshift pass
 TESTS PASSED [358s on mysystem]

File: pkg/kubernetes/scripts/projects.js [hunk 16]
```diff
@@ -406,6 +418,29 @@
...
+                    kind: 'User',
+                    name: toName(currentUser),
+                };
+                //if system:admin then return
+                if (currentUser && currentUser.groups &&
+                    currentUser.groups.indexOf("system:cluster-admins") !== -1) {
+                    return $q.when({ });
+                }
+                var oc_roles = getRegistryRoles(currentUser, project);
+                if (oc_roles.indexOf(registryAdmin) !== -1) {
```
Comment (sub-mod) [inline pkg/kubernetes/scripts/projects.js]: @aweiteka I check if i need to add registry-admin role

**Comment by sub-mod** [comment]
@stefwalter I don think the test failures are due to my code.

File: pkg/kubernetes/scripts/kube-client-cockpit.js [hunk 17]
```diff
@@ -1070,6 +1071,14 @@
                         var req = new CockpitKubeRequest("GET", "/oapi", "", options)
                             .then(function() {
                                 settings.flavor = "openshift";
+                                loader.load("/oapi/v1/users/~")
```
Comment (sub-mod) [inline pkg/kubernetes/scripts/kube-client-cockpit.js]: @petervo adding the code only for openshift.

**Comment by sub-mod** [comment]
ok 1 testBasic (__main__.TestTuned)
ok 1 testConnect (__main__.TestConnection)
ok 5 testProjectUsers (__main__.TestRegistry)
All testcase pass now
@petervo

File: pkg/kubernetes/scripts/projects.js (hunk 15 above)
Comment (aweiteka) [inline pkg/kubernetes/scripts/projects.js]: Yes, this is the approach we discussed. Looks good.

Comment (aweiteka) [inline pkg/kubernetes/scripts/projects.js]: > IMO we need to add a project template which will not create a 'admin' policy. ok ?

Once this is merged and we know more about how this will be deployed we will adjust as needed. With this work we have less need to have special project templates. This is a really positive direction.

**Comment by aweiteka** [comment]
LGTM
